x-pack/dockerlogbeat/pipereader/reader.go

```diff
@@ -0,0 +1,101 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package pipereader
+
+import (
+	"context"
+	"encoding/binary"
+	"io"
+	"syscall"
+
+	"github.com/docker/engine/api/types/plugins/logdriver"
+	"github.com/gogo/protobuf/proto"
+	"github.com/pkg/errors"
+	"github.com/tonistiigi/fifo"
+)
+
+// PipeReader reads from the FIFO pipe we get from the docker container
+type PipeReader struct {
+	fifoPipe    io.ReadWriteCloser
+	byteOrder   binary.ByteOrder
+	lenFrameBuf []byte
+	bodyBuf     []byte
+	maxSize     int
+}
+
+// NewReaderFromPath creates a new FIFO pipe reader from a docker log pipe location
+func NewReaderFromPath(file string) (*PipeReader, error) {
+	inputFile, err := fifo.OpenFifo(context.Background(), file, syscall.O_RDONLY, 0700)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error opening logger fifo: %q", file)
+	}
+
+	return &PipeReader{fifoPipe: inputFile, byteOrder: binary.BigEndian, lenFrameBuf: make([]byte, 4), bodyBuf: nil, maxSize: 2e6}, nil
+}
```

Comment:
Consider to introduce a constructor with a io.ReadCloser parameter instead of opening the file. Inject dependency, instead of creating them.

This also simplifies unit testing (e.g. introduce some randomzied testing using quickcheck or fuzzing). Unit test could be driven from fifo, pipe, or simple buffer. The protobuf writers only require an io.Writer. This can be used to prepare a buffer that is used with the reader afterwards. Simplified this means the property `x == read(write(x))` must hold.

For convencience we can still keep this constructor.